core/services/llo/keyring.go

```diff
@@ -0,0 +1,78 @@
+package llo
+
+import (
+	"fmt"
+
+	"github.com/smartcontractkit/libocr/offchainreporting2/types"
+	"github.com/smartcontractkit/libocr/offchainreporting2plus/ocr3types"
+	ocrtypes "github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+
+	commontypes "github.com/smartcontractkit/chainlink-common/pkg/types"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+)
+
+type LLOOnchainKeyring ocr3types.OnchainKeyring[commontypes.LLOReportInfo]
+
+var _ LLOOnchainKeyring = &onchainKeyring{}
+
+type Key interface {
+	Sign3(digest ocrtypes.ConfigDigest, seqNr uint64, r ocrtypes.Report) (signature []byte, err error)
+	Verify3(publicKey ocrtypes.OnchainPublicKey, cd ocrtypes.ConfigDigest, seqNr uint64, r ocrtypes.Report, signature []byte) bool
+	PublicKey() ocrtypes.OnchainPublicKey
+	MaxSignatureLength() int
+}
+
+type onchainKeyring struct {
+	lggr logger.Logger
+	keys map[string]Key
+}
+
+func NewOnchainKeyring(lggr logger.Logger, keys map[string]Key) LLOOnchainKeyring {
+	return &onchainKeyring{
+		lggr.Named("OnchainKeyring"), keys,
+	}
+}
+
+func (okr *onchainKeyring) PublicKey() types.OnchainPublicKey {
+	// All public keys combined
+	var pk []byte
+	for _, k := range okr.keys {
+		pk = append(pk, k.PublicKey()...)
+	}
+	return pk
+}
+
+func (okr *onchainKeyring) MaxSignatureLength() (n int) {
+	// Needs to be max of all chain sigs
+	for _, k := range okr.keys {
+		n += k.MaxSignatureLength()
+	}
+	return
+}
+
+func (okr *onchainKeyring) Sign(digest types.ConfigDigest, seqNr uint64, r ocr3types.ReportWithInfo[commontypes.LLOReportInfo]) (signature []byte, err error) {
+	rf := string(r.Info.ReportFormat)
+	// HACK: sign/verify JSON payloads with EVM keys for now, this makes
+	// debugging and testing easier
+	if rf == "json" {
```

Comment:
add a check that ReportFormat is one of the thing you expect? what happens if it's not json and not evm?

Reply:
It's handled on line 60, if we don't have a matching report format in the map it returns error

Reply:
https://github.com/smartcontractkit/chainlink-common/pull/352